### PR TITLE
chore(docs): documenting reference, how-to and explainer on the usage of oracles

### LINKED ISSUE #3971: Document the `#[oracle]` attribute
### Problem

The `#[foreign_call]` attribute has been surfaced in multiple discussions as a feature for fetching information outside of the Noir program, but we don't currently have written explanations of how it is used.

### Happy Case

Document how is `#[foreign_call]` used in the Noir docs with examples.

### Alternatives Considered

_No response_

### Additional Context

Non-exhausive / potentially inaccurate list of reference PRs I came across:
- https://github.com/noir-lang/noir/pull/3294
- https://github.com/noir-lang/noir/pull/3506
- https://github.com/noir-lang/noir/pull/3764
- https://github.com/noir-lang/noir/pull/3902

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by signorecello**
# Description

This PR adds some more in-depth documentation on oracles, what they are and how to use them.

## Problem\*

Closes #3971

Oracles can now be resolved with JSON RPC calls, which is a powerful feature for developers. This PR attempts to provide some information on how they can be part of a development flow

## Summary\*

- [ ] Adds the nargo commands for oracles
- [ ] Adds an explainer on what are oracles and what can you do with them
- [ ] Adds a how-to guide on using oracles
- [ ] Adds a tutorial on a small project using oracles

### CODE REVIEW

**Comment by signorecello** [comment]
Superseeds #3975

**Comment by signorecello** [comment]
I think this is in a good spot for reviewing

File: docs/docs/how_to/how-to-oracles.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+For example, if your circuit expects the host machine to provide CPU pseudo-randomness, you could simply pass it as the `foreignCallHandler`. You don't strictly need to create an RPC server to serve pseudo-randomness, as you may as well get it directly in your app:
+
+```js
+const foreignCallHandler = (name, inputs) => crypto.randomBytes(16) // etc
+
+await noir.generateFinalProof(inputs, foreignCallHandler)
+```
+
+As one can see, in NoirJS, the [`foreignCallHandler`](../reference/NoirJS/noir_js/type-aliases/ForeignCallHandler.md) function simply means "a callback function that returns a value of type [`ForeignCallOutput`](../reference/NoirJS/noir_js/type-aliases/ForeignCallOutput.md). It doesn't have to be an RPC call like in the case for Nargo.
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:196]: These two links can't be resolved when i run the docs locally.

File: docs/docs/explainers/explainer-oracle.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+An example usage for Oracles is proving something on-chain. For example, proving that the ETH-USDC quote was below a certain target at a certain block time. Or even making more complex proofs like proving the ownership of an NFT as an anonymous login method.
+
+Another interesting use case is to defer expensive calculations to be made outside of the circuit, and then constraining the result. This is no different from the example in the [unconstrained page](../noir/concepts//unconstrained.md).
+
+In short, anything that can be constrained in a circuit but needs to be fetched from an outside source is a great candidate to be used in oracles.
+
+## Constraining oracles
+
+Just like in Matrix, Oracles are powerful. But with great power, comes great responsibility. Just because you're using them in a Noir program, that doesn't mean they're true. Noir has no superpowers: if you want to prove that Portugal won the Euro Cup 2016, you're still relying on potentially untrusted information.
```
Comment (critesjosh) [inline docs/docs/explainers/explainer-oracle.md:32]:
```suggestion
Just like in The Matrix, Oracles are powerful. But with great power, comes great responsibility. Just because you're using them in a Noir program doesn't mean they're true. Noir has no superpowers. If you want to prove that Portugal won the Euro Cup 2016, you're still relying on potentially untrusted information.
```

File: docs/docs/explainers/explainer-oracle.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+To give a concrete example, Alice wants to login to the [NounsDAO](https://nouns.wtf/) forum with her username "noir_nouner" by proving she owns a noun without revealing her ethereum address. Her Noir program could have a oracle call like this:
+
+```rust
+#[oracle(getNoun)]
+unconstrained fn get_noun(address: u32) -> Field
```
Comment (critesjosh) [inline docs/docs/explainers/explainer-oracle.md:38]: An ethereum address can't fit in u32

```suggestion
unconstrained fn get_noun(address: Field) -> Field
```

File: docs/docs/explainers/explainer-oracle.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+```
+
+This oracle could naively resolve with the number of Nouns she possesses. However, it is useless as a trusted source, as the oracle could resolve to anything Alice wants. In order to make this oracle call actually useful, Alice would need to constrain the response from the oracle, by proving her address and the noun count belongs to the state tree of the contract.
+
+In short, **Oracles don't prove anything. Your Noir program does.**
```
Comment (critesjosh) [inline docs/docs/explainers/explainer-oracle.md:43]: Well said!

File: docs/docs/explainers/explainer-oracle.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+:::danger
+
+If you don't constrain the return of your oracle, you could be clearly opening an attack vector on your Noir program. Make double-triple sure that the return of an oracle call is constrained!
+
+:::
+
+## How to use Oracles
+
+Currently, Nargo will resolve oracles by making JSON RPC calls, which means an RPC node needs to be running. NoirJS, on the other hand, can provide whatever call handler as long as it matches the expected types.
```
Comment (critesjosh) [inline docs/docs/explainers/explainer-oracle.md:53]: > NoirJS, on the other hand, can provide whatever call handler as long as it matches the expected types.

What does this mean exactly? Can you explain a bit more?

File: docs/docs/how_to/how-to-oracles.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+  - Noir Programming
+  - Oracles
+  - Nargo
+  - NoirJS
+  - JSON RPC Server
+  - Foreign Call Handlers
+sidebar_position: 1
+---
+
+This guide shows you how to use oracles in your Noir program. For the sake of clarity, it is assumed that:
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:14]:
```suggestion
This guide shows you how to use oracles in your Noir program. For the sake of clarity, it is assumes that:
```

File: docs/docs/how_to/how-to-oracles.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+- You have read the [explainer on Oracles](../explainers/explainer-oracle.md) and are comfortable with the concept
+- You already have your own NoirJS and/or Nargo app. If you don't, you can use the [vite-hardhat starter](https://github.com/noir-lang/noir-starter/tree/main/vite-hardhat) as a boilerplate (for example, by using its [devcontainer in codespaces](./using-devcontainers.mdx))
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:17]:
```suggestion
- You already have your own NoirJS and/or Nargo app. If you don't, you can use the [vite-hardhat starter](https://github.com/noir-lang/noir-starter/tree/main/vite-hardhat) as a boilerplate (for example, by using its [devcontainer in codespaces](./using-devcontainers.mdx)).
```

File: docs/docs/how_to/how-to-oracles.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+  - Nargo
+  - NoirJS
+  - JSON RPC Server
+  - Foreign Call Handlers
+sidebar_position: 1
+---
+
+This guide shows you how to use oracles in your Noir program. For the sake of clarity, it is assumed that:
+
+- You have read the [explainer on Oracles](../explainers/explainer-oracle.md) and are comfortable with the concept
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:16]: for consistency
```suggestion
- You have read the [explainer on Oracles](../explainers/explainer-oracle.md) and are comfortable with the concept.
```

File: docs/docs/how_to/how-to-oracles.md [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+- You already have your own NoirJS and/or Nargo app. If you don't, you can use the [vite-hardhat starter](https://github.com/noir-lang/noir-starter/tree/main/vite-hardhat) as a boilerplate (for example, by using its [devcontainer in codespaces](./using-devcontainers.mdx))
+- You understand the concept of a JSON-RPC server. Visit the [JSON-RPC website](https://www.jsonrpc.org/) if you need a refresher.
+- You are comfortable with server-side javascript. Will skip any details on installing Node, packages, etc, so as to keep the guide short and straight to the point.
+
+For reference, you can find the snippets used in this tutorial on the [Aztec DevRel Repository](https://github.com/AztecProtocol/dev-rel/tree/main/how_to_oracles/code-snippets/how-to-oracles)
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:21]:
```suggestion
For reference, you can find the snippets used in this tutorial on the [Aztec DevRel Repository](https://github.com/AztecProtocol/dev-rel/tree/main/how_to_oracles/code-snippets/how-to-oracles).
```

File: docs/docs/how_to/how-to-oracles.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+## Rundown
+
+This guide has 3 major steps:
+
+1. How to modify our circuit to make use of oracle calls as unconstrained functions.
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:27]:
```suggestion
1. How to modify our circuit to make use of oracle calls as unconstrained functions
```

File: docs/docs/how_to/how-to-oracles.md [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+2. How to write a JSON RPC Server to resolve these oracle calls with Nargo
+3. How to use them in Nargo and how to provide a custom resolver in NoirJS
+
+## Step 1 - Modify your circuit
+
+An oracle is defined in a circuit by defining two methods:
+
+- An unconstrained method - This tells the Noir compiler that it is executing a Brillig block.
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:35]: Should you add a link to "Brillig" for those that are not familiar with the term to learn more?

File: docs/docs/how_to/how-to-oracles.md [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+In this example, we're wrapping our oracle function in a unconstrained method, and decorating it with `oracle(getSquared)`. We can then call the unconstrained function as we would call any other function:
+
+```rust
+fn main(input: Field) {
+    let squared = get_squared(input);
+}
+```
+
+In the next section, we will make this `getSquared` be a method of the RPC server Noir will use.
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:57]: It'd be useful to call out that the RPC server will use the method defined in the decorator, not the name of the Noir function. This confused me for more than a minute.

```suggestion
In the next section, we will make this `getSquared` (defined on the `squared` decorator) be a method of the RPC server Noir will use.

File: docs/docs/how_to/how-to-oracles.md [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+unconstrained fn get_squared(number: Field) -> Field {
+    squared(number)
+}
+```
+
+In this example, we're wrapping our oracle function in a unconstrained method, and decorating it with `oracle(getSquared)`. We can then call the unconstrained function as we would call any other function:
+
+```rust
+fn main(input: Field) {
+    let squared = get_squared(input);
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:53]: Is it required to have the `squared` function wrapped in the `get_squared` function? We can't just call `squared` directly in the `main` fn? if so, why?

File: docs/docs/how_to/how-to-oracles.md [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+await noir.generateFinalProof(inputs, foreignCallHandler)
+```
+
+As one can see, in NoirJS, the [`foreignCallHandler`](../reference/NoirJS/noir_js/type-aliases/ForeignCallHandler.md) function simply means "a callback function that returns a value of type [`ForeignCallOutput`](../reference/NoirJS/noir_js/type-aliases/ForeignCallOutput.md). It doesn't have to be an RPC call like in the case for Nargo.
+
+:::tip
+
+Does this mean you don't have to write an RPC server like in [Step #2](#step-2---write-an-rpc-server)?
+
+You don't technically have to, but then how would you run `nargo test` or `nargo prove`? To use both `Nargo` and `NoirJS` in your development flow, you will most certainly have to write a JSON RPC server anyway.
```
Comment (critesjosh) [inline docs/docs/how_to/how-to-oracles.md:202]: less :slightly_smiling_face: 
```suggestion
You don't technically have to, but then how would you run `nargo test` or `nargo prove`? To use both `Nargo` and `NoirJS` in your development flow, you will have to write a JSON RPC server.
```

File: docs/docs/how_to/how-to-oracles.md (hunk 1 above)
Comment (critesjosh) [reply docs/docs/how_to/how-to-oracles.md:196]: Do i need to run a command or something locally so they are generated before starting the docusaurus server? If so, it'd be useful to add the command to `yarn start` in the docs package.json

File: docs/docs/noir/concepts/oracles.md [hunk 15]
```diff
@@ -0,0 +1,23 @@
+---
+title: Understanding Oracles in Noir - A Comprehensive Guide
```
Comment (critesjosh) [inline docs/docs/noir/concepts/oracles.md:2]: Shorten to match the style of the rest of the pages in this section 
image


```suggestion
title: Oracles
```

File: docs/docs/noir/concepts/oracles.md [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+  - RPC Calls
+  - Unconstrained Functions
+  - Programming
+  - Blockchain
+sidebar_position: 6
+---
+
+Noir has support for Oracles via RPC calls. This means Noir will make an RPC call and use the return value for proof generation.
+
+Since Oracles are not resolved by Noir, they are obviously [`unconstrained` functions](./unconstrained.md)
```
Comment (critesjosh) [inline docs/docs/noir/concepts/oracles.md:16]: less :slightly_smiling_face: 
```suggestion
Since Oracles are not resolved by Noir, they are [`unconstrained` functions](./unconstrained.md)
```

File: docs/docs/noir/concepts/oracles.md
Comment (critesjosh) [inline docs/docs/noir/concepts/oracles.md:1]: I like how short and to the point this page is, but if this is the first page about Oracles that someone lands on, it leads to many questions. Should we link to the Oracles how to guide and/or explainer page?

File: docs/docs/reference/NoirJS/noir_js/index.md [hunk 17]
```diff
@@ -26,7 +26,7 @@
 | :------ | :------ |
 | [and](functions/and.md) | Performs a bitwise AND operation between `lhs` and `rhs` |
 | [blake2s256](functions/blake2s256.md) | Calculates the Blake2s256 hash of the input bytes |
-| [ecdsa\_secp256k1\_verify](functions/ecdsa_secp256k1_verify.md) | Verifies a ECDSA signature over the secp256k1 curve. |
+| [ecdsa\_secp256k1\_verify](functions/ecdsa_secp256k1_verify.md) | Calculates the Blake2s256 hash of the input bytes and represents these as a single field element. |
```
Comment (critesjosh) [inline docs/docs/reference/NoirJS/noir_js/index.md:29]: This seems wrong.

File: docs/docs/reference/NoirJS/noir_js/functions/ecdsa_secp256k1_verify.md [hunk 18]
```diff
@@ -8,6 +8,7 @@ ecdsa_secp256k1_verify(
    signature): boolean
 ```
 
+Calculates the Blake2s256 hash of the input bytes and represents these as a single field element.
```
Comment (critesjosh) [inline docs/docs/reference/NoirJS/noir_js/functions/ecdsa_secp256k1_verify.md:11]: This is out of place as well.

File: docs/docs/reference/NoirJS/backend_barretenberg/index.md
Comment (critesjosh) [inline docs/docs/reference/NoirJS/backend_barretenberg/index.md:1]: These changes are unrelated to oracles and the other autogen code (blake hashes in ECDSA verify doc) is wrong, so just checking that it's correct.

**Review by critesjosh — CHANGES_REQUESTED** [review]

File: docs/docs/how_to/how-to-oracles.md (hunk 11 above)
Comment (Savio-Sou) [reply docs/docs/how_to/how-to-oracles.md:35]:
```suggestion
- An unconstrained method - This tells the compiler that it is executing an [unconstrained functions](../noir/concepts//unconstrained.md).
```

File: docs/docs/how_to/how-to-oracles.md [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+nargo test --oracle-resolver http://localhost:5555
+```
+
+This tells `nargo` to use your RPC Server URL whenever it finds an oracle decorator.
+
+## Step 4 - Usage with NoirJS
+
+In a JS environment, an RPC server is not strictly necessary, as you may want to resolve your oracles without needing any JSON call at all. NoirJS simply expects that you pass a callback function when you generate proofs, and that callback function can be anything.
+
+For example, if your circuit expects the host machine to provide CPU pseudo-randomness, you could simply pass it as the `foreignCallHandler`. You don't strictly need to create an RPC server to serve pseudo-randomness, as you may as well get it directly in your app:
```
Comment (Savio-Sou) [inline docs/docs/how_to/how-to-oracles.md:188]:
```suggestion
For example, if your Noir program expects the host machine to provide CPU pseudo-randomness, you could simply pass it as the `foreignCallHandler`. You don't strictly need to create an RPC server to serve pseudo-randomness, as you may as well get it directly in your app:
```

File: docs/docs/how_to/how-to-oracles.md [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+
+:::tip
+
+Does this mean you don't have to write an RPC server like in [Step #2](#step-2---write-an-rpc-server)?
+
+You don't technically have to, but then how would you run `nargo test` or `nargo prove`? To use both `Nargo` and `NoirJS` in your development flow, you will most certainly have to write a JSON RPC server anyway.
+
+:::
+
+In this case, let's make `foreignCallHandler` call the JSON RPC Server we created in [Step #2](#step-2---write-an-rpc-server), by making it a JSON RPC Client. For example, using the same `getSquared` circuit in [Step #1](#step-1---modify-your-circuit) (comments in the code):
```
Comment (Savio-Sou) [inline docs/docs/how_to/how-to-oracles.md:206]:
```suggestion
In this case, let's make `foreignCallHandler` call the JSON RPC Server we created in [Step #2](#step-2---write-an-rpc-server), by making it a JSON RPC Client.

For example, using the same `getSquared` program in [Step #1](#step-1---modify-your-noir-program) (comments in the code):
```

File: docs/docs/reference/NoirJS/noir_js/index.md (hunk 17 above)
Comment (Savio-Sou) [reply docs/docs/reference/NoirJS/noir_js/index.md:29]:
```suggestion
| [ecdsa_secp256k1_verify](functions/ecdsa_secp256k1_verify.md) | Verifies a ECDSA signature over the secp256k1 curve. |
```

File: docs/docs/reference/NoirJS/noir_js/index.md [hunk 21]
```diff
@@ -26,7 +26,7 @@
 | :------ | :------ |
 | [and](functions/and.md) | Performs a bitwise AND operation between `lhs` and `rhs` |
 | [blake2s256](functions/blake2s256.md) | Calculates the Blake2s256 hash of the input bytes |
```
Comment (Savio-Sou) [inline docs/docs/reference/NoirJS/noir_js/index.md:28]:
```suggestion
| [blake2s256](functions/blake2s256.md) | Calculates the Blake2s256 hash of the input bytes and represents these as a single field element. |
```

File: docs/docs/reference/NoirJS/noir_js/index.md [hunk 22]
```diff
@@ -26,7 +26,7 @@
...
-| [ecdsa\_secp256k1\_verify](functions/ecdsa_secp256k1_verify.md) | Verifies a ECDSA signature over the secp256k1 curve. |
+| [ecdsa\_secp256k1\_verify](functions/ecdsa_secp256k1_verify.md) | Calculates the Blake2s256 hash of the input bytes and represents these as a single field element. |
 | [ecdsa\_secp256r1\_verify](functions/ecdsa_secp256r1_verify.md) | Verifies a ECDSA signature over the secp256r1 curve. |
```
Comment (Savio-Sou) [inline docs/docs/reference/NoirJS/noir_js/index.md:30]:
```suggestion
| [ecdsa_secp256r1_verify](functions/ecdsa_secp256r1_verify.md) | Verifies a ECDSA signature over the secp256r1 curve. |
```

File: docs/docs/reference/NoirJS/noir_js/functions/ecdsa_secp256k1_verify.md (hunk 18 above)
Comment (Savio-Sou) [reply docs/docs/reference/NoirJS/noir_js/functions/ecdsa_secp256k1_verify.md:11]:
```suggestion
```

**Review by Savio-Sou — COMMENTED** [review]
Minor nits.

File: docs/docs/how_to/how-to-oracles.md (hunk 1 above)
Comment (signorecello) [reply docs/docs/how_to/how-to-oracles.md:196]: Yes these point to autogenerated docs but they should be generated with yarn start... I'll investigate a bit

File: docs/docs/how_to/how-to-oracles.md (hunk 11 above)
Comment (signorecello) [reply docs/docs/how_to/how-to-oracles.md:35]: Perfect @Savio-Sou but I still see so much "brillig" everywhere, like debugger, etc. It's a discussion for another time

File: docs/docs/how_to/how-to-oracles.md (hunk 13 above)
Comment (signorecello) [reply docs/docs/how_to/how-to-oracles.md:53]: Just followed the tests here, apparently you need to wrap the oracle in an unconstrained function. I get a warning message telling me that, too.

Not amazing but I would guess the decorator runs a macro or something that transforms the function into an oracle.

File: docs/docs/how_to/how-to-oracles.md (hunk 14 above)
Comment (signorecello) [reply docs/docs/how_to/how-to-oracles.md:202]: Lessssss 😈

(we should put that as #2 rule of devrel - #1 will always be "tag a DevRel")

File: docs/docs/noir/concepts/oracles.md (hunk 15 above)
Comment (signorecello) [reply docs/docs/noir/concepts/oracles.md:2]: Shit, I always finish by asking ChatGPT to write the frontmatter because of SEO, and forgot to change the title 😂

File: docs/docs/noir/concepts/oracles.md
Comment (signorecello) [reply docs/docs/noir/concepts/oracles.md:1]: That's the main reason I would like to have a tutorial. But there are resource constraints here in terms of devrel time as we spoke offline.

To answer you question, I would point to the explainer, at least for now. But this is part of the dev journey work I'll do next week...

File: docs/docs/reference/NoirJS/noir_js/index.md (hunk 17 above)
Comment (signorecello) [reply docs/docs/reference/NoirJS/noir_js/index.md:29]: I don't even know why this file was committed. It is autogenerated and will be autogenerated at build time. Could be worth checking if the typedoc decorator itself is wrong, I'll have a look

File: docs/docs/how_to/how-to-oracles.md (hunk 1 above)
Comment (signorecello) [reply docs/docs/how_to/how-to-oracles.md:196]: Fixed. Issue happened because I was rebuilding _just_ the docs, thus it was getting wrong info from the updated TS build from the monorepo.

For future reference: to avoid this, don't be a lazy ass like me and go run `yarn build` from the root folder.

File: docs/docs/reference/NoirJS/noir_js/index.md (hunk 17 above)
Comment (signorecello) [reply docs/docs/reference/NoirJS/noir_js/index.md]: Fixed as explained above 👍

File: docs/docs/reference/NoirJS/backend_barretenberg/index.md
Comment (signorecello) [reply docs/docs/reference/NoirJS/backend_barretenberg/index.md:1]: Fixed as explained above 👍

File: docs/docs/how_to/how-to-oracles.md [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+---
+
+This guide shows you how to use oracles in your Noir program. For the sake of clarity, it assumes that:
+
+- You have read the [explainer on Oracles](../explainers/explainer-oracle.md) and are comfortable with the concept.
+- You have a Noir program to add oracles to. You can create one using the [vite-hardhat starter](https://github.com/noir-lang/noir-starter/tree/main/vite-hardhat) as a boilerplate.
+- You understand the concept of a JSON-RPC server. Visit the [JSON-RPC website](https://www.jsonrpc.org/) if you need a refresher.
+- You are comfortable with server-side JavaScript (e.g. Node.js, managing packages, etc.).
+
+For reference, you can find the snippets used in this tutorial on the [Aztec DevRel Repository](https://github.com/AztecProtocol/dev-rel/tree/main/how_to_oracles/code-snippets/how-to-oracles).
```
Comment (Savio-Sou) [inline docs/docs/how_to/how-to-oracles.md:21]: Link to code is dead and could use an update.

**Review by Savio-Sou — APPROVED** [review]

File: docs/docs/how_to/how-to-oracles.md (hunk 23 above)
Comment (signorecello) [reply docs/docs/how_to/how-to-oracles.md:21]: This is intended, as it points to an unmerged PR in the devrel repo

File: docs/docs/how_to/how-to-oracles.md [hunk 24]
```diff
@@ -38,32 +38,32 @@ An oracle is defined in a Noir program by defining two methods:
...
 
 As explained in the [Oracle Explainer](../explainers/explainer-oracle.md), this `main` function is unsafe unless you constrain its return value. For example:
 
 ```rust
-fn main(input: Field, target: Field) {
-    let squared = get_squared(input);
-    assert(squared as u64 == target as u64); // <---- constrain the return of an oracle!
+fn main(input: Field) {
+    let sqrt = get_sqrt(input);
+    assert(sqrt[0].pow_32(2) as u64 == input as u64); // <---- constrain the return of an oracle!
```
Comment (sirasistant) [inline docs/docs/how_to/how-to-oracles.md:66]: I think this has to be just sqrt.pow_32!

**Review by sirasistant — APPROVED** [review]
Just one typo!

File: docs/docs/how_to/how-to-oracles.md (hunk 23 above)
Comment (Savio-Sou) [reply docs/docs/how_to/how-to-oracles.md:21]: Thanks Ze. For the record, the PR is https://github.com/AztecProtocol/dev-rel/pull/118.

**Review by critesjosh — APPROVED** [review]
:+1:
